File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/property/GitBranchConfigurationProperty.java
package net.nemerosa.ontrack.extension.git.property;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.experimental.Wither;
import net.nemerosa.ontrack.model.structure.ServiceConfiguration;

@Data
public class GitBranchConfigurationProperty {

    /**
     * Git branch
     */
    private final String branch;

    /**
     * Build link
     */
    private final ServiceConfiguration buildCommitLink;

    /**
     * Build overriding policy when synchronizing
     */
    private final boolean override;

    /**
     * Interval in minutes for build/tag synchronization
     */
    private final int buildTagInterval;

}

File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/property/GitBranchConfigurationProperty.kt
package net.nemerosa.ontrack.extension.git.property

import net.nemerosa.ontrack.model.annotations.APIDescription
import net.nemerosa.ontrack.model.docs.DocumentationField
import net.nemerosa.ontrack.model.structure.PropertySearchArguments
import net.nemerosa.ontrack.model.structure.ServiceConfiguration

class GitBranchConfigurationProperty(

    /**
     * Git branch or pull request ID
     */
    @APIDescription("Git branch or pull request ID")
    val branch: String,

    /**
     * Build link
     */
    @DocumentationField
    @APIDescription("How builds are linked to their Git commit")
    val buildCommitLink: ServiceConfiguration?,

    /**
     * Build overriding policy when synchronizing
     */
    @APIDescription("Build overriding policy when synchronizing")
    val override: Boolean,

    /**
     * Interval in minutes for build/tag synchronization
     */
    @APIDescription("Interval in minutes for build/tag synchronization")
    val buildTagInterval: Int

) {
    companion object {
        fun getSearchArguments(scmName: String): PropertySearchArguments? =
            if (scmName.isNotBlank()) {
                PropertySearchArguments(
                    jsonContext = null,
                    jsonCriteria = "pp.json->>'branch' = :scmName",
                    criteriaParams = mapOf("scmName" to scmName)
                )
            } else {
                null
            }
    }
}
